### PR TITLE
Membership syncing

### PR DESCRIPTION
**Pull request by Cadair**
This is a WIP to implement full membership syncing with slack. It will eventually support reacting to join and leave events on slack and then also creating ghosts for all the slack users in the room at startup.

### CODE REVIEW

File: src/BridgedRoom.ts [hunk 1]
```diff
@@ -176,6 +176,40 @@ export class BridgedRoom {
...
+        }
+        const ghost = await this.main.getGhostForSlack(user_id, this.team.domain, this.team.id);
+        return ghost.intent.join(this.matrixRoomId);
+    }
+
+    private async getSlackChannelMembership() {
+        if (!this.botClient || !this.slackChannelId) {
+            return;
+        }
+        // TODO: Handle paging here
```
Comment (Half-Shot) [inline src/BridgedRoom.ts]: It's almost certainly true that if botClient exists, we have a team. I would urge you to just return early if `this.teams` isn't defined. And then I would try store the bot user_id in a `const` at the start and use that const in the filter.

File: src/BridgedRoom.ts [hunk 2]
```diff
@@ -176,6 +176,34 @@ export class BridgedRoom {
...
 
         this.dirty = true;
+
+        // This almost certainly isn't the best place for this
+        this.joinAllSlackUsers();
+    }
+
+    public async joinAllSlackUsers() {
+        const members = await this.getSlackChannelMembership();
+        if (!members) return;
```
Comment (Half-Shot) [inline src/BridgedRoom.ts]:
```suggestion
```

Don't bother doing this, a `Promise.all([].map(...))` will do the right thing.

File: src/BridgedRoom.ts [hunk 3]
```diff
@@ -176,6 +176,34 @@ export class BridgedRoom {
         this.puppetOwner = opts.puppet_owner;
 
         this.dirty = true;
+
+        // This almost certainly isn't the best place for this
+        this.joinAllSlackUsers();
```
Comment (Half-Shot) [inline src/BridgedRoom.ts]: I would call `joinAllSlackUsers` during startup (queued, obviously), and in `actionLink`. Calling it in the constructor not only means we are not awaiting it, but also will cause unintended side effects by doing this in a constructor (e.g. if we do our cached rooms thing we talked about in #265)

File: src/BridgedRoom.ts (hunk 2 above)
Comment (Cadair) [reply src/BridgedRoom.ts]: The typescript compiler disagrees with you.

File: src/SlackEventHandler.ts [hunk 4]
```diff
@@ -144,6 +151,26 @@ export class SlackEventHandler extends BaseSlackHandler {
         }
     }
 
+    protected async handleMemberJoined(event: ISlackEvent, teamId: string) {
```
Comment (Half-Shot) [inline src/SlackEventHandler.ts]: This should be `private`

File: src/SlackEventHandler.ts [hunk 5]
```diff
@@ -144,6 +151,26 @@ export class SlackEventHandler extends BaseSlackHandler {
...
+        const teamDomain = (await this.main.getTeamDomainForMessage(undefined, teamId)).toLowerCase();
+        if ( !event.user) {
+            throw Error("No user given for event.");
+        }
+        const ghost = await this.main.getGhostForSlack(event.user, teamDomain, teamId);
+        const room = this.main.rooms.getBySlackChannelId(event.channel) as BridgedRoom;
+        await ghost.intent.join(room.MatrixRoomId);
+    }
+
+    protected async handleMemberParted(event: ISlackEvent, teamId: string) {
```
Comment (Half-Shot) [inline src/SlackEventHandler.ts]: This should be `private`

File: package.json [hunk 6]
```diff
@@ -6,7 +6,7 @@
   "scripts": {
     "postinstall": "npm run build",
     "start": "node ./lib/app.js",
-    "build": "tsc",
+    "build": "tsc -sourcemap",
```
Comment (Half-Shot) [inline package.json]: put this in `tsconfig.json`?

File: src/BridgedRoom.ts (hunk 2 above)
Comment (Half-Shot) [reply src/BridgedRoom.ts]: Oh bother, because it can be undefined? Fair

File: src/BridgedRoom.ts [hunk 7]
```diff
@@ -178,6 +178,32 @@ export class BridgedRoom {
...
+
+    private async joinUserToRoom(user_id) {
+        if (!this.team) return;
+        const ghost = await this.main.getGhostForSlack(user_id, this.team.domain, this.team.id);
+        return ghost.intent.join(this.matrixRoomId);
+    }
+
+    private async getSlackChannelMembership() {
+        if (!this.team || !this.botClient) return;
+        const our_user_id = this.team.user_id;
```
Comment (Half-Shot) [inline src/BridgedRoom.ts]: `camelCase`, not `snake_cake`

File: src/Main.ts [hunk 8]
```diff
@@ -274,13 +274,13 @@ export class Main {
         return `${(hs.media_url || hs.url)}/_matrix/media/r0/download/${mxcUrl.substring("mxc://".length)}`;
     }
 
-    public async getTeamDomainForMessage(message: any, teamId?: string) {
-        if (message.team_domain) {
+    public async getTeamDomainForMessage(message?: any, teamId?: string) {
```
Comment (Half-Shot) [inline src/Main.ts]: This feel like the wrong approach, why are we calling getTeamDomain on unknown messages?

File: src/Main.ts [hunk 9]
```diff
@@ -796,16 +796,22 @@ export class Main {
         log.info(`Found ${entries.length} room entries in store`);
         const joinedRooms = await roomListPromise;
         i = 0;
-        await Promise.all(entries.map(async (entry) => {
+        const rooms = await Promise.all(entries.map(async (entry) => {
             i++;
             log.info(`[${i}/${entries.length}] Loading room entry ${entry.matrix_id}`);
             try {
-                await this.startupLoadRoomEntry(entry, joinedRooms, teamClients);
+                return await this.startupLoadRoomEntry(entry, joinedRooms, teamClients);
```
Comment (Half-Shot) [inline src/Main.ts]: return isn't needed

File: src/Main.ts [hunk 10]
```diff
@@ -796,16 +796,22 @@ export class Main {
...
             } catch (ex) {
                 log.error(`Failed to load entry ${entry.matrix_id}, exception thrown`, ex);
             }
         }));
 
+        const roomMembersPromises = new PQueue({concurrency: STARTUP_TEAM_INIT_CONCURRENCY});
+        for (const room of rooms) {
+            if (!room) return;
+            roomMembersPromises.add(async () => room.joinAllSlackUsers());
```
Comment (Half-Shot) [inline src/Main.ts]: async isn't needed here, is it?

File: src/Main.ts [hunk 11]
```diff
@@ -796,16 +796,22 @@ export class Main {
...
-                await this.startupLoadRoomEntry(entry, joinedRooms, teamClients);
+                return await this.startupLoadRoomEntry(entry, joinedRooms, teamClients);
             } catch (ex) {
                 log.error(`Failed to load entry ${entry.matrix_id}, exception thrown`, ex);
             }
         }));
 
+        const roomMembersPromises = new PQueue({concurrency: STARTUP_TEAM_INIT_CONCURRENCY});
+        for (const room of rooms) {
+            if (!room) return;
```
Comment (Half-Shot) [inline src/Main.ts]:
```suggestion
            if (!room) continue;
```

File: src/Main.ts (hunk 8 above)
Comment (Cadair) [reply src/Main.ts]: I should probably rename the method, it's more that we want to get the domain from the teamId.

**Comment by Half-Shot** [comment]
(Sorry, I may have done this in #332) Thanks for kicking us off though!

**Comment by Cadair** [comment]
Did you add initial sync?

**Comment by Half-Shot** [comment]
I did indeed.

**Pull request closed without merging**